Admin store item and collection management

SheepTester opened the pull request:
<!-- Fill in all spots surrounded by brackets and confirm all check boxes after confirming completion of the tasks -->

# Info

Closes #123. Depends on #94 

<!-- If there is no issue for this pull request yet, please create one or
delete this line if the pull request is for a very minor tweak. -->

I added store item and collection edit and create pages under `/store/item/new`, `/store/item/[uuid]/edit`, `/store/collection/new`, and `/store/collection/[uuid]/edit`. Their code is based on the event edit/create page. I also added edit and create buttons to the public storefront pages, but admins can hide these with an option in the admin panel that stores this preference in a cookie.

This also adds a `Draggable` component for reordering lists of elements. You just give it an array and a render function, and it'll make each element draggable. It works best when every element is the same size.

<!-- What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context. -->

## Changes

In addition to the changes made above,

- I changed apiRequests/Responses from .d.ts -> .ts because it was obscuring some type errors/unknown types
- I enforced file sizes for profile pictures and store merch images
- The collection and item pages return a 404 error for missing items instead of redirecting to the store home page
- In "view all items" mode, the storefront collection sliders subtly link to their collection pages

Changes after first reviews:

- I also updated the rest of the codebase to prefer `reportError` for API errors and `cond ? jsx : null` instead of `cond && jsx`
- ~~The service worker now fetches from the network first instead of loading stale data from the cache~~
- Moved extra arguments of `withAccessType` into an options object because it became unclear what they mean

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) <!-- Edit this type of change if you select this -->

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [x] on the live deployment preview on Desktop.
- [x] on the live deployment preview on Mobile.
- [ ] I have added new Cypress tests that are passing.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented any new functions in `/src/lib/*` and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

Edit buttons on various parts of the website

Storefront: collections view | Storefront: all items view | Item page
--- | --- |---
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/5b8200c2-d9a8-41c3-94f7-6c07ce48534e)|![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/53e96a47-4402-4e24-b5e3-ef492655ceb7)|![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/9a4fd03f-37f0-4520-ab38-220e88ee680a)

New collection page:
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/befbf826-b647-4be6-89f8-6a890cb43f63)

Edit item page:
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/ffcc3386-16fa-4257-a085-c9ffa0af5643)

Reordering images and options:

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/1cb8864b-03a5-47e3-86e5-977c1a05ff8f)
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/e2d5df1e-462d-432d-9b6e-8827b5d840ac)

WishingWell13 committed:
Rough outline of SizeSelector

WishingWell13 committed:
Created a file to represent the item page

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Edited config file to be nested under "Store" header

WishingWell13 committed:
Size Toggle V1, Styling Only

Size toggle styling implemented

On selected text goes blue

On hover border goes blue

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
State of SizeToggle now passed up to the item page

Made a temporary edit to the title of the page that displays the current size selected as testing

WishingWell13 committed:
Updated styling of size selector to match Figma

https://www.figma.com/file/cgbeW6vrqgxW0i1vpbUn1C/Store-(Redesign-2023)?type=design&node-id=2028-16321&mode=design&t=nbIfHwdT8aeDdhsd-0

WishingWell13 committed:
Part 1 of AddCartButton styling

Got text color to change depending on if in stock or not (is currently hard coded, not synced to backend)

Button animation still exists, need to make it flat

Text changes based on if in stock or not

Still need to do styling for button if item is out of stock

WishingWell13 committed:
Out of Stock Button and "Please Select Size"

Updated button theming to change based on if in stock or not

A size is now not selected by default, message requesting user to select a size added.

WishingWell13 committed:
Fixed issue with AddCartButton props breaking

From last update, where made it possible for size to be undefined.

WishingWell13 committed:
Began formatting for item header and item page

WishingWell13 committed:
Store Component now linked to API, price added

Price now displays. No discount display yet

TODO: Modify size buttons to be dynamic depending on how many options there are.

WishingWell13 committed:
Size Selection Buttons Now Dynamically Generated

farisashai committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Buy ability based on lifetime/monthly limits, added item picture

TODO: Might be a good idea to change the "add to cart" message depending on if you have reached lifetime limits

WishingWell13 committed:
Cleaned up some old comments

WishingWell13 committed:
Moved Navbar to top of page

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Resolved linting error (unused var)

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/lib/api/StoreAPI.ts

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Resolved merge error with calendar

And API type error

WishingWell13 committed:
Minor package changes

alexzhang1618 committed:
Fix package.json

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Make compatible with multiple store gallery images

No carousel yet, uses default photo as photo shown

WishingWell13 committed:
Moved item header to be above sizing buttons

farisashai committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Centered the body of the item page.

WishingWell13 committed:
Fixed merge error by updating name of file path

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/pages/store/item/[uuid].tsx

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Update src/pages/store/item/[uuid].tsx

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Update src/pages/store/item/[uuid].tsx

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Update .vscode/settings.json

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/styles/pages/store/item.module.scss

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/components/store/AddCartButton/index.tsx

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Addressed PR comments, Add to Cart button behavior edit

Updated Add to Cart button to also say "remove from cart"

Quantity Selector now disappear if you can't add to cart or size is not selected.

Add to Cart button disappear if a size is not selected.

Removed various comments

Removed useEffect

WishingWell13 committed:
Updated diamond component in Item Header

Diamond component displays cost

WishingWell13 committed:
Add To Cart button disappears when UserPurchaseLimit reached

WishingWell13 committed:
AddCartButton maxCanBuy saftey checks and styling

Added an option for 0 quantity if user has reached purchase limit

WishingWell13 committed:
Renamed styling files, AddToCart Button Behavior Change

Add To Cart button now displays when user has reached purchase limit, but is greyed out and cannot be clicked

WishingWell13 committed:
Added border back to disabled AddToCart Button

WishingWell13 committed:
Added border back to cart button, cleaned up unused vars

WishingWell13 committed:
Hid toggle when maxCanBuy = 1

Also lifted quantity of items up to uuid.tsx

WishingWell13 committed:
Removed unused comments

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

farisashai committed:
Merge branch 'main' into andy/StoreItemPage

WishingWell13 committed:
Removed eslint disable lines

WishingWell13 committed:
Removed title attribute from buy button

https://github.com/acmucsd/membership-portal-ui-v2/pull/94#discussion_r1443580814

WishingWell13 committed:
Removed trailing spaces from button text

https://github.com/acmucsd/membership-portal-ui-v2/pull/94#discussion_r1443581247

WishingWell13 committed:
Hide price when size is not selected

Also removed mounted checks from `SizeSelector`

importing only types from UUID

WishingWell13 committed:
Added back other types accidently deleted from last commit

WishingWell13 committed:
Delete blue-diamond.svg

WishingWell13 committed:
Basic mobile formatting implemented

Still need to sync it with Figma

SheepTester committed:
Add edit buttons to store page

SheepTester committed:
Merge remote-tracking branch 'origin/andy/StoreItemPage' into sean/admin-store-item

SheepTester committed:
Copy event edit page for store item edit page

SheepTester committed:
item details form

SheepTester committed:
Set item collection, preview store as member

SheepTester committed:
Add create buttons to storefront page

SheepTester committed:
Collection edit page

SheepTester committed:
API for creating/deleting store item photos/options

SheepTester committed:
Fix making changes to collections

WishingWell13 committed:
Fixed image clipping issue, reduced image gap on mobile

WishingWell13 committed:
Removed vertical scrollbar from mobile view

Used scrollbar hide, may need to look into why is actually overflowing later.

WishingWell13 committed:
Removed unnecessary CSS

SheepTester committed:
item options table

SheepTester committed:
Add multiple options

SheepTester committed:
fix type error

SheepTester committed:
Merge remote-tracking branch 'origin/main' into andy/StoreItemPage

SheepTester committed:
Rename store/collections.module.scss to be consistent with other changes in PR

SheepTester committed:
Merge remote-tracking branch 'origin/andy/StoreItemPage' into sean/admin-store-item

SheepTester committed:
Fix duplicate `getCollection` definitions

SheepTester committed:
Make API .d.ts files normal TypeScript files so it actually reports errors

SheepTester committed:
Upload photos

SheepTester committed:
Enforce limits

SheepTester committed:
PermissionService.allUserTypes is no longer a function

SheepTester committed:
Make edit and create button components and add them to collection page

SheepTester committed:
Add edit button to item page, show 404 when item/collection DNE

SheepTester committed:
Show hidden/archived icons

SheepTester commented:
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/ecaa8b99-ef57-41c4-a237-9872ff47a372)

SheepTester committed:
Make the edit forms look nicer

WishingWell13 committed:
Fixed a linting error with the options dropdown

WishingWell13 committed:
Shrunk some UI items for mobile version of site

WishingWell13 committed:
Set upper limit on quantity of items can buy

WishingWell13 committed:
Deleted unnecessary comments

WishingWell13 committed:
Update src/components/store/SizeSelector/index.tsx

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Fixed various linting errors, removed comments

Also updated gaps

WishingWell13 committed:
Decreased font size of dropdown text

SheepTester committed:
fix item cards' edit button

SheepTester committed:
Merge remote-tracking branch 'origin/andy/StoreItemPage' into sean/admin-store-item

WishingWell13 committed:
Merge branch 'main' into andy/StoreItemPage

SheepTester committed:
Overengineered Draggable component

SheepTester committed:
drag handle design adjustments

SheepTester committed:
Some input validation

SheepTester committed:
Save preview as member option in cookie rather than URL

SheepTester committed:
My JavaScript is too future for Vercel 😔

SheepTester committed:
Make design look more acceptable on mobile

farisashai reviewed (changes requested):
When you click "View store as member" on the store home page, can we trigger a refresh so that it actually switches to the user preview because right now nothing will change until refreshing. The alternative is store client state in the component which works too but is a little more overhead

farisashai reviewed (changes requested):
when the checkbox "Preview store as member" is checked, can you update the text in the button "Manage Store Merchandise" to "View Merch Store" so the actions of the checkbox and button below it are synchronized with the actual behavior that will take place

farisashai reviewed (changes requested):
i'm going to leave some comments here that are only relevant for the store item page too and review it all at once

- when you enter an item page with variants, the first option that isn't out of stock should be selected by default to prevent layout shift from the price appearing after selecting an option

farisashai reviewed (changes requested):
if you flip the "preview as member" checkbox -> go to store -> return to admin page and repeat this cycle while never refreshing the page, i'm seeing that the checkbox value can easily end up out of sync with the cookie value - i wonder if this is because the page is using outdated props and not re-rendering?

farisashai reviewed (changes requested):
idk how feasible this is but after you click "Create Collection" and get re-routed to the modify collection page, clicking back on your browser should ideally take you to the store home page instead of the create collection page because the form has already been submitted and shouldn't be created again

farisashai reviewed (changes requested):
on the "Create Store Item" page, what does quantity mean in the options? i think you might want to indicate what these fields are setting a little bit better. also it looks like you're missing a place to define variants such as S, M, L, XL as options

farisashai reviewed (changes requested):
when a collection or store item is created or edited successfully, i think you should take the user to the collection or item page where they can see their changes visually to receive confirmation that everything was set up right - i don't think someone would immediately want to modify the collection/item on the same form immediately after creating it (unless there is a mistake which they can see by viewing the item)

farisashai reviewed (changes requested):
might be nice to list which collection this item is a part of on the item page if that data exists easily

farisashai reviewed (changes requested):
you should enforce validation that the lifetime limit >= monthly limit

farisashai commented on src/components/admin/store/CollectionDetailsForm/index.tsx:
```diff
@@ -0,0 +1,182 @@
+import EventDetailsFormItem from '@/components/admin/event/EventDetailsFormItem';
```
can you import this as a named import or does that introduce a cyclical dependency

alexzhang1618 replied:
Let's refactor `EventDetailsFormItem` into a common component named `DetailsFormItem` or `FormItem` since the file seems pretty generic and it's being used in two places now

farisashai commented on src/components/admin/store/CollectionDetailsForm/index.tsx:
```diff
@@ -0,0 +1,182 @@
+import { useState } from 'react';
+import { SubmitHandler, useForm } from 'react-hook-form';
+import { BsArrowRight } from 'react-icons/bs';
+import style from '../ItemDetailsForm/style.module.scss';
```
seeking to learn why we're importing a css file outside the current folder

SheepTester replied:
They have the same CSS, so I didn't want to copy it. Ideally I would rearrange the files, but I'm not sure how. I'll find a way

farisashai commented on src/components/admin/store/CollectionDetailsForm/index.tsx:
```diff
@@ -0,0 +1,182 @@
+  const initialValues: FormValues = {
+    title: defaultData.title ?? '',
+    // ACM blue
+    themeColorHex: defaultData.themeColorHex ?? '#62b0ff',
```
wondering if there's a way to pull the css var directly from a sass sheet for this

other option is just referencing a css variable if that works 
```suggestion
    themeColorHex: defaultData.themeColorHex ?? 'var(whatever-blue-variable)',
```

farisashai commented on src/components/admin/store/CollectionDetailsForm/index.tsx:
```diff
@@ -0,0 +1,182 @@
+  const createCollection: SubmitHandler<FormValues> = async formData => {
+    setLoading(true);
+
+    try {
```
can you move any client-side API calls to a StoreAdminManager file so we can isolate out any business logic relating to making the api call in that function and keep any stateful component updates in the callbacks to make it more readable and separate concerns like if the api schema changes for example we know where to look

farisashai commented on src/components/admin/store/CollectionDetailsForm/index.tsx:
```diff
@@ -0,0 +1,182 @@
+
+    const uuid = defaultData.uuid ?? '';
+
+    try {
```
same as above, move to a manager function

farisashai commented on src/components/store/CollectionSlider/index.tsx:
```diff
@@ -1,32 +1,60 @@
 import { Carousel } from '@/components/common';
+import CreateButton from '@/components/store/CreateButton';
```
import all three of these from @/components/store in one line as named imports

SheepTester replied:
since this is also a store component, there will be a dependency circle because `@/components/store` already depends on `@/components/CollectionSlider`

SheepTester commented:
Cookies seem very finicky; I'll have to do some more testing with them (unsure how client vs server cookies work, and when they're synched). Originally preview mode was kept in a URL parameter rather than as a local preference

> idk how feasible this is but after you click "Create Collection" and get re-routed to the modify collection page, clicking back on your browser should ideally take you to the store home page instead of the create collection page because the form has already been submitted and shouldn't be created again

Oh this is real easy, I can just use router replace instead of push



> on the "Create Store Item" page, what does quantity mean in the options? i think you might want to indicate what these fields are setting a little bit better. also it looks like you're missing a place to define variants such as S, M, L, XL as options

I can change quantity to Quantity Available

The variant is the leftmost column that appears when there are two or more options

farisashai commented on src/components/store/ItemCard/index.tsx:
```diff
@@ -26,25 +27,34 @@ type ItemCardProps = (StoreItemOptions | CollectionOptions) & CommonOptions;
+        </div>
+        <div className={styles.details}>
+          <p className={styles.title}>{title}</p>
+          {'description' in props && <p>{props.description}</p>}
```
im not 100% sure but i feel like it might be better to write this like `<p>{props?.description ?? ""}</p>` so react only has to update the text and we don't have the entire element rendering from scratch

SheepTester replied:
I would rather the empty paragraph element not be there. Also, `description` for `ItemCard` shouldn't change often (and the way it's currently used, it won't change without a page reload)

farisashai commented on src/components/store/SizeSelector/index.tsx:
```diff
@@ -11,21 +11,23 @@ interface SizeSelectorProps {
 const SizeSelector = ({ currSize, options, onSizeChange }: SizeSelectorProps) => {
   const myID = useId();
```
this might be a bit late but can someone explain to me why we have to define IDs and can't rely on component state to track everything it needs

SheepTester replied:
This comes from Andy's PR, but IDs are used so that inputs and labels can be flattened (`<input id={id} /> <label htmlFor={id}>...</label>` instead of `<label><input /></label>`). This allows the CSS to do `input:checked + label` etc and make `label` (which is not a replaced element so it's easier to style) to be what the user sees as the radio button, whilst still preserving built-in browser functionality.

I suppose we could use `label:has(:checked)` now, but that CSS feature feels uncomfortably new

We can use JavaScript to track the checked state of each radio button, but it's a lot more code than just relying on CSS.

farisashai commented on src/pages/profile/edit.tsx:
```diff
@@ -15,25 +15,14 @@ import withAccessType from '@/lib/hoc/withAccessType';
 import Link from 'next/link';
 import { FormEvent, useEffect, useId, useMemo, useState } from 'react';
 
-function reportError(title: string, error: unknown) {
```
nice improvement! if you could confirm everywhere is using this function, that would be really good

farisashai commented on src/pages/profile/edit.tsx:
```diff
@@ -15,25 +15,14 @@ import withAccessType from '@/lib/hoc/withAccessType';
+import { capitalize, fixUrl, getProfilePicture, reportError } from '@/lib/utils';
 import DownloadIcon from '@/public/assets/icons/download-icon.svg';
 import DropdownIcon from '@/public/assets/icons/dropdown-arrow-1.svg';
 import styles from '@/styles/pages/profile/edit.module.scss';
```
```suggestion
import styles from '@/styles/pages/EditProfile.module.scss';
```
This isn't caused by your PR but can we flatten all of the page level CSS sheets to one folder by renaming them and fixing the imports because there's not really any reason to nest them in folders like we do with pages to define the routing structure.

farisashai commented on src/pages/store/collection/[uuid].tsx:
```diff
@@ -11,35 +11,53 @@ import styles from '@/styles/pages/StoreCollectionPage.module.scss';
       <div className={styles.header}>
         <Typography variant="h1/bold" component="h1">
           {title}
+          {canManageStore && <EditButton type="collection" uuid={uuid} />}
```
i might have missed this syntax existing in a few previous PRs but we actually have a standard on this for conditional rendering where we should always write it as 
 ```suggestion
          {canManageStore ? <EditButton type="collection" uuid={uuid} /> : null}
```

the bug this is avoiding is where you have a check such as 
```
const quantity = 0;
{quantity && <MyComponent />}
```
you would expect this not to render because the conditional is falsy however javascript shortcircuiting will still return the first value in this case so react will render a plaintext 0 on the screen. by writing it with the suggested syntax, this bug is not possible since the conditional is never the returned value

farisashai commented on src/pages/store/index.tsx:
```diff
@@ -1,5 +1,14 @@
+  ItemCard,
+  Navbar,
+} from '@/components/store';
+import CreateButton from '@/components/store/CreateButton';
```
import in group above

farisashai reviewed (changes requested):
excellent work mr sheeptester you are a frontend god and win the trophy for fastest and largest prs in history with a very clean implementation 🏆 i'm gonna be honest i stopped reading every line at some point but i think i did a decent code quality review - hopefully someone else can do a full run through of the diff. either way, for a feature of this size, we should have at least 3 different people stress test the preview deployment for edge cases and UI bugs that can be addressed before merging it so i would hold off on merging until we have enough validation in that sense.

two more comments i didn't mention yet
- i'm seeing a small bug on iOS when you drag an item image in edit UI where the box shadow leaves a trail across the page, i can try to screen record it later if you can't reproduce it
- on the admin actions homepage, the buttons text overflows on small screen sizes, if you can unhardcode the height? on these, or fix that UI issue, that would be excellent

farisashai commented on src/lib/types/enums.ts:
```diff
@@ -1,6 +1,7 @@
 export enum CookieType {
   USER = 'USER',
   ACCESS_TOKEN = 'ACCESS_TOKEN',
+  PREVIEW = 'PREVIEW',
```
this variable isn't very descriptive currently as to what kind of browser state it's storing, we should rename to something like "USER_PREVIEW_ENABELD" or similar

alexzhang1618 commented on src/components/admin/store/CollectionDetailsForm/index.tsx:
```diff
@@ -0,0 +1,182 @@
+          <input
+            type="text"
+            id="title"
+            placeholder="ACM Cafe"
```
nit: let's change the placeholder to something more relevant

alexzhang1618 replied:
Here and in ItemDetailsForm

alexzhang1618 commented on src/components/admin/store/ItemDetailsForm/index.tsx:
```diff
@@ -0,0 +1,571 @@
+                </td>
+              </tr>
+            </tfoot>
+          </table>
```
Is this worth moving into a separate file to make the code more concise?

alexzhang1618 commented on src/components/store/ItemCard/index.tsx:
```diff
@@ -26,25 +27,34 @@ type ItemCardProps = (StoreItemOptions | CollectionOptions) & CommonOptions;
+          {'description' in props && <p>{props.description}</p>}
+          {'cost' in props && (
+            <p className={styles.cost}>
+              <Diamonds count={props.cost} />
```
Can we also add the discount value here on Diamonds? I've noticed it's not being displayed when it should

alexzhang1618 commented on src/pages/admin/index.tsx:
```diff
@@ -26,14 +28,27 @@ const AdminPage = ({ user: { accessType } }: AdminProps) => {
+            CookieService.setClientCookie(
+              CookieType.PREVIEW,
+              e.currentTarget.checked ? 'member' : 'admin'
+            )
```
seeking to learn: should we set the client cookie or the server cookie here? I see in gSSP that we're picking from the server cookie

SheepTester replied:
looks like faris answered this on discord

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/d6973708-b715-471f-aba6-065950565f89)

alexzhang1618 commented on src/pages/store/collection/[uuid].tsx:
```diff
@@ -11,35 +11,53 @@ import styles from '@/styles/pages/StoreCollectionPage.module.scss';
+  collection: { title, description, items = [], archived },
+  previewPublic,
 }: CollectionProps) => {
+  const canManageStore = PermissionService.canEditMerchItems.includes(accessType) && !previewPublic;
```
nit: let's rewrite this into something more descriptive like storeAdminVisible

alexzhang1618 commented on src/pages/store/index.tsx:
```diff
@@ -9,18 +18,42 @@ import { getDefaultMerchItemPhoto } from '@/lib/utils';
+
   const [helpOpen, setHelpOpen] = useState(false);
 
+  const canManageStore = PermissionService.canEditMerchItems.includes(accessType) && !previewPublic;
```
again, rewrite to something like canViewStoreAdmin?

alexzhang1618 commented on src/pages/store/index.tsx:
```diff
@@ -31,37 +64,71 @@ const StoreHomePage = ({ user: { credits }, view, collections }: HomePageProps)
+              type="button"
+              className={styles.viewToggle}
+              onClick={() => {
+                CookieService.setClientCookie(CookieType.PREVIEW, 'member');
```
Same question here with client vs. server cookies

alexzhang1618 reviewed (commented):
Functionality is 🔥! 
There's some weird cookie stuff going on where with setting the view in settings and moving to the store doesn't reveal the new view unless you refresh. 
Do you think it'll work better if we use exclusively client cookies to check if we should display admin tools?

WishingWell13 committed:
New styling changes to adjust spacing

IN PROGRESS

WishingWell13 committed:
Locked width of description and fixed line breaks

Also adjusted title/paragraph sizes and weight

Also centered the webpage

WishingWell13 committed:
Title now properly wraps with very long titles

Also deleted some unused styling classes

Also adjusted one column width threshold.

WishingWell13 committed:
Made padding a little bigger for right column

Also fixed bug where price would cause an extra gap in object header, even when not visible

WishingWell13 committed:
Removed in stock / out of stock messages

Also bumped the "add to cart" button up a little bit

WishingWell13 committed:
@media now uses the media breakpoint variables

WishingWell13 committed:
Renamed variables and adjusted white space behavior

WishingWell13 committed:
Update src/components/store/AddCartButton/index.tsx

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Update src/components/store/AddCartButton/style.module.scss

Co-authored-by: Sean <ussa.say@gmail.com>

WishingWell13 committed:
Fixed some gap inconsistencies

WishingWell13 committed:
Update settings.json

WishingWell13 committed:
Update settings.json

WishingWell13 committed:
Made SizeSelector title change based on metadata

WishingWell13 committed:
Fixed bug where add to cart button no longer worked

Was due to comparison for current option not being properly updated with new system (needed a `.value`)

Also made image container always square.

WishingWell13 committed:
Condensed imports in `uuid`, removed "Remove from Cart" text

Note that clicking "Add to Cart" again removes item from cart.

SheepTester committed:
Merge remote-tracking branch 'origin/andy/StoreItemPage' into sean/admin-store-item

SheepTester committed:
Fuss with store component imports to avoid dependency cycles

SheepTester committed:
Flatten profile/edit.module.scss

SheepTester committed:
Replace all `condition && jsx` with `condition ? jsx : null` to conform with style guide

Per https://github.com/acmucsd/membership-portal-ui-v2/pull/124#discussion_r1464544092

SheepTester committed:
Update placeholder text for name; fix going back after creating item going to create page

The collection/item title placeholder text has been made more relevant

After creating an item, it redirects you to the edit page so when you click back, it doesn't go to an empty create item form anymore. Same with collection

SheepTester committed:
Restructure details form component files

SheepTester committed:
Quantity -> Quantity available in table, redirect to public page after creating item/collection, fix admin buttons on mobile

WishingWell13 committed:
Before attempting to remove currOption

I think I can collapse it into selectedOption

SheepTester committed:
rename cookie type PREVIEW -> USER_PREVIEW_ENABLED

WishingWell13 committed:
Made the "please select an option" text to show key value

So instead of always saying "size" or "option" it will show "color" if the options are colors

SheepTester committed:
Add discount and out of stock to item card

SheepTester committed:
canManageStore -> storeAdminVisible

WishingWell13 committed:
Update src/components/store/AddCartButton/style.module.scss

Co-authored-by: Sean <ussa.say@gmail.com>

SheepTester committed:
Refactor store api stuff to manager, options table to component

WishingWell13 committed:
Changed AddCartButton to CartOptionsGroup

Also various small bug fixes from PR

- Made the options `input` tags not selectable

- Removed unecessary CSS

WishingWell13 committed:
Merge branch 'andy/StoreItemPage' of https://github.com/acmucsd/membership-portal-ui-v2 into andy/StoreItemPage

WishingWell13 committed:
Removed unnecessary properties, simplified CSS

SheepTester committed:
Refactor all managers to use `reportError`

SheepTester committed:
Change Manage Store Merchandise button to View Merch Store when preview is checked

SheepTester committed:
Merge remote-tracking branch 'origin/andy/StoreItemPage' into sean/admin-store-item

SheepTester committed:
Debugging cookie desync issue

SheepTester committed:
how about collection page

SheepTester committed:
Change service worker strategy to network first

This fixes the admin preview store as member cookie being out of sync, but it also means slower page loads for the user

SheepTester committed:
Require that monthlyLimit <= lifetimeLimit

SheepTester commented:
I messed with the next-pwa settings and now the admin preview cookie desync issues are gone :crab:

alexzhang1618 commented on next.config.js:
```diff
@@ -1,8 +1,26 @@
 const env = process.env.NODE_ENV;
 const isDevelopment = env !== 'production';
 
+const runtimeCaching = require('next-pwa/cache');
+
+// By default, next-pwa serves from the cache while updating the cache with a
+// network request. This results in faster page loads, but the data will be
+// stale, most notable when checking into an event (membership points will be
+// out of date) or enabling "Preview store as member" on the admin page. Note
+// that this only applies to static routes (e.g. /store but not /u/[uuid]).
+//
+// This makes it fetch data from the server (then falling back to the cache if
+// the user is offline). This will result in slower page loads, but the data
+// will always be up-to-date.
+// https://github.com/vercel/next.js/discussions/52024#discussioncomment-6325542
+const nextData = runtimeCaching.find(entry => entry.options.cacheName === 'next-data');
+if (nextData) {
+  nextData.handler = 'NetworkFirst';
+}
+
 const withPWA = require('next-pwa')({
   dest: 'public',
+  runtimeCaching,
```
Can we try merging main in and disabling caching for specific pages instead of site-wide?

It should just be 
`withAccessType(... true)` at the bottom of the page where gSSP is being exported

SheepTester replied:
I don't think `disableCaching` works. Clicking on "View store as member" doesn't do anything until you reload the page or click again, like it was originally.

farisashai commented on src/components/admin/store/DetailsForm/CollectionDetailsForm.tsx:
```diff
@@ -0,0 +1,172 @@
+  const createCollection: SubmitHandler<FormValues> = async formData => {
+    setLoading(true);
+
+    const uuid = await AdminStoreManager.createNewCollection(token, formData);
```
how are you error handling here if an api error gets thrown?

farisashai commented on src/components/admin/store/DetailsForm/CollectionDetailsForm.tsx:
```diff
@@ -0,0 +1,172 @@
+
+    const uuid = defaultData.uuid ?? '';
+
+    if (await AdminStoreManager.editCollection(token, uuid, formData)) {
```
can we do async operations on their own line and just save it to a variable to use here for better readability

SheepTester committed:
Merge remote-tracking branch 'origin/main' into sean/admin-store-item

big conflict because we squash and merge prs 😔

SheepTester committed:
Minor code quality adjustments from store item PR

SheepTester committed:
Fix store slider scrollbars in Chrome 121+

alexzhang1618 commented on src/components/admin/store/DetailsForm/ItemDetailsForm.tsx:
```diff
@@ -0,0 +1,361 @@
+          } else if (reason !== null) {
+            showToast('This image format is not supported.');
+          }
+        }}
```
It feels like we generally send the same error for every Cropper component, thoughts on bringing this inside the component and not allowing it to be customized?

SheepTester committed:
Only disable caching for store pages

Also change `withAccessType` to use an options object so it's more clear what the parameters do

alexzhang1618 reviewed (approved):
we move 🚢

SheepTester committed:
Handle showing error toasts in Cropper

SheepTester committed:
export scss variable to default collection color

SheepTester committed:
minor nitfixes

farisashai reviewed (approved):
small comments but looks good

SheepTester committed:
Move `reportError` out of admin store manager to make error handling more clear

SheepTester committed:
make spacing of code look more consistent

SheepTester committed:
Revert back to global NetworkFirst, remove `disableCaching`
